### PR TITLE
frontend: Add Single Sign-on to Toolkit

### PR DESCRIPTION
**Pull request by malexw**
Updates the login page from `alexw/coral-login` to dynamically add and hide the username / password form and a couple of single sign-on buttons depending on the backend's enabled auth strategies.

**AI Description**

This PR introduces changes to the authentication system, including the addition of new authentication strategies and updates to the existing ones. 

- The `JWT_SECRET_KEY` has been renamed to `AUTH_SECRET_KEY` in the `.env-template` file and the `auth.guide.md` documentation.
- The `auth.guide.md` file has been updated to reflect the changes in the authentication process.
- The `ENABLED_AUTH_STRATEGIES` list in `auth.py` now includes `BasicAuthentication`, `GoogleOAuth`, and `OpenIDConnect`.
- The `src/backend/main.py` file has been updated to import the `asyncio` and `os` modules, and the `get_strategy_endpoints` function has been added to the list of imports from `auth.py`.
- The `src/backend/main.py` file now includes the `SessionMiddleware` middleware, which is required for temporary OAuth state storage in the session.
- The `src/backend/routers/auth.py` file has been updated to include the `get_client_id` and `get_authorization_endpoint` methods, and the `login` function has been modified to use the new authentication strategies.
- The `src/backend/services/auth/jwt.py` file now uses the `AUTH_SECRET_KEY` environment variable instead of `JWT_SECRET_KEY`.
- The `src/backend/services/auth/strategies/base.py` file has been updated to include the `get_client_id`, `get_authorization_endpoint`, and `get_endpoints` methods, and the `authorize` method has been added to the `BaseOAuthStrategy` class.
- The `src/backend/services/auth/strategies/google_oauth.py` and `src/backend/services/auth/strategies/oidc.py` files have been modified to include the `get_client_id`, `get_authorization_endpoint`, and `get_endpoints` methods, and the `authorize` method has been added to the `GoogleOAuth` and `OpenIDConnect` classes.
- The `src/interfaces/coral_web/src/components/Welcome/OidcSSOButton.tsx` file has been added to provide a button for OIDC sign-up and login.
- The `src/interfaces/coral_web/src/hooks/authConfig.ts` file has been updated to include the `useServerAuthStrategies` hook and return the list of authentication strategies.
- The `src/interfaces/coral_web/src/hooks/githubAuthRoute.ts` file has been removed.
- The `src/interfaces/coral_web/src/hooks/oidcAuthRoute.ts` file has been updated to use the `useAuthConfig` hook and include the `handleOidcAuth` function.
- The `src/interfaces/coral_web/src/hooks/session.ts` file now includes the `googleSSOMutation` and `oidcSSOMutation` mutations for handling Google and OIDC SSO, respectively.
- The `src/interfaces/coral_web/src/pages/auth/[strategy].tsx` and `src/interfaces/coral_web/src/pages/auth/complete.tsx` files have been added to handle the callback from OAuth providers and forward the request to the backend.
- The `src/interfaces/coral_web/src/pages/login.tsx` file has been updated to include the `useOidcAuthRoute` hook and the `useMemo` and `useAuthConfig` functions. The `useGoogleAuthRoute` hook has been removed, and the `OidcSSOButton` component has been added for OIDC sign-up and login.

### CODE REVIEW

File: src/backend/config/auth.py [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 # Add Auth strategy classes here to enable them
 # Ex: [BasicAuthentication]
-ENABLED_AUTH_STRATEGIES = []
+ENABLED_AUTH_STRATEGIES = [BasicAuthentication, GoogleOAuth, OpenIDConnect]
```
Comment (tianjing-li) [inline src/backend/config/auth.py:5]: Let's not to commit these

File: src/backend/services/auth/strategies/google_oauth.py [hunk 2]
```diff
@@ -19,47 +20,40 @@ class GoogleOAuth(BaseOAuthStrategy):
...
         try:
-            settings = GoogleOAuthSettings()
-            self.oauth = OAuth()
-            self.oauth.register(
-                name="google",
-                server_metadata_url="https://accounts.google.com/.well-known/openid-configuration",
-                client_id=settings.google_client_id,
-                client_secret=settings.google_client_secret,
-                client_kwargs={"scope": "openid email profile"},
+            self.settings = GoogleOAuthSettings()
```
[the remark this answers is not in the document]
Comment (tianjing-li) [reply src/backend/services/auth/strategies/google_oauth.py:27]: This looks good to me

**Review by tianjing-li — APPROVED** [review]
Changes look good to me!

File: src/backend/routers/auth.py [hunk 3]
```diff
@@ -132,7 +136,8 @@ async def oidc_authenticate(request: Request, session: DBSessionDep):
     """
     strategy_name = OpenIDConnect.NAME
 
-    return await authenticate(request, session, strategy_name)
+    # TODO..
```
Comment (BeatrixCohere) [inline src/backend/routers/auth.py:139]: Can you give more details here

File: src/interfaces/coral_web/src/hooks/oidcAuthRoute.ts [hunk 4]
```diff
@@ -1,33 +1,37 @@
...
-  // const authConfig = useAuthConfig();
-  const authConfig = {
-    login: {
-      googleClientId: 'googleClientId',
-    },
-    baseUrl: 'baseUrl',
-  };
+export const useOidcAuthRoute = () => {
+  const authConfig = useAuthConfig();
+  // const googleStrategy: { strategy: string, clientId: string } = authConfig.login.find((strategy) => strategy.strategy === 'Google') || {
```
Comment (BeatrixCohere) [inline src/interfaces/coral_web/src/hooks/oidcAuthRoute.ts:14]: Should this still be here?

**Review by BeatrixCohere — APPROVED** [review]

File: src/interfaces/coral_web/src/hooks/oidcAuthRoute.ts (hunk 4 above)
Comment (malexw) [reply src/interfaces/coral_web/src/hooks/oidcAuthRoute.ts]: Whoops, thanks!

**Pull request merged**